docs/design.md

```diff
@@ -0,0 +1,194 @@
+out for the controller, and do the work of keeping the automation and
+policy objects in the cluster up to date with what's in the git repo
+(one way to do that is to keep them in the git repo and let them be
+synced; but in general, the workloads are not going to run in the same
+place as the automation, so it would be fiddly to keep these in the
+same place).
+
+A design in the other direction would be to expect annotations,
+similar to those used by Flux v1, on workloads to be automated. The
+controller would interpret those to determine which image repositories
+and policies are needed.
+
+This might be tricky when the automation is managed by a different
+team to that in charge of the application configuration -- in that
+scenario, the annotations would have to be carefully applied after the
+fact (perhaps with a kustomization), which may as well mean the
+annotations are kept in a separate file.
+
+## Easy (Flux v1) mode
+
+This design includes an "easy mode" for simple deployments, which
+mimics closely the behaviour of the automation in Flux v1. Creating an
+`ImageAutomation` resource that analyses the annotations in git
+repository will give much the same results as in Flux v1.
+
+## Extension points
+
+**Actions other than UpdateJob**
+
+As a way in to other systems, e.g., Tekton, the `ImageAutomation` job
+could specify an action other than creating an `UpdateJob` (for
+Tekton, this might be creating a `TaskRun`). The impedence here is how
+to package the information about what to update -- this will need to
+be thought through for each such integration.
+
+**Supply a different image to UpdateJob**
+
+If the way this design makes commits doesn't suit, then a different
+image can be supplied to the `UpdateJob` resource created by
+automation.
+
+**Writing a controller**
+
+As a last resort, it's possible to write a controller that reacts to
+`ImagePolicy` changes in its own way, or analyses the files in the git
+repository according to some other scheme. It can hook back into the
+rest of the design by using `UpdateJob` resources to accomplish
+whatever it calculates needs to be done.
+
+## Open questions
+
+**How does the controller authenticate with image repositories?**
+
+Flux v1 picks up credentials from the cluster, e.g., by looking at the
+image pull secrets for a workload while it's looking for images to
+scan. It also tries to sense when it can get credentials from the
+environment, e.g., by asking GCP for credentials. That basically
+works, modulo getting some details right and keeping up with platform
+changes.
+
+In this design, it would be possible to adapt the Flux v1 scheme, more
+or less. For an `ImageRepository`, the controller could look through
+all the workloads to find one using the image, and use its image pull
+secrets. And, it could try to sense when it can use ambient
+authorisation.
+
+It might be better to at least start by being explicit, though, and
```

Comment:
As a starting point, this seems better. One aspect to consider is RBAC, will we allow the controller to read any Deployment/DaemonSet/Pod-container type (and Secrets) by default?

Reply:
Probably not by default, by in one configuration yes.